fairmq/MemoryResourceTools.h

```diff
@@ -0,0 +1,117 @@
+ *                  copied verbatim in the file "LICENSE" *
+ ********************************************************************************/
+
+/// @brief Tools for interfacing containers to the transport via polymorphic
+/// allocators
+///
+/// @author Mikolaj Krzewicki, mkrzewic@cern.ch
+
+#include <fairmq/FairMQTransportFactory.h>
+#include <fairmq/MemoryResources.h>
+
+namespace fair {
+namespace mq {
+
+using ByteSpectatorAllocator = SpectatorAllocator<fair::mq::byte>;
+using BytePmrAllocator = boost::container::pmr::polymorphic_allocator<fair::mq::byte>;
+
+//_________________________________________________________________________________________________
+// return the message associated with the container or nullptr if it does not
+// make sense (e.g. when
+// we are just watching an existing message or when the container is not using
+// FairMQMemoryResource
+// as backend).
+template<typename ContainerT>
+// typename std::enable_if<
+//    std::is_base_of<
+//        boost::container::pmr::polymorphic_allocator<typename
+//        ContainerT::value_type>,
+//        typename ContainerT::allocator_type>::value == true,
+//    FairMQMessagePtr>::type
+FairMQMessagePtr getMessage(ContainerT &&container_, FairMQMemoryResource *targetResource = nullptr)
+{
+    auto container = std::move(container_);
+    auto alloc = container.get_allocator();
+
+    auto resource = dynamic_cast<FairMQMemoryResource *>(alloc.resource());
+    if (!resource && !targetResource) {
+        throw std::runtime_error("Neither the container or target resource specified");
+    }
+    size_t containerSizeBytes = container.size() * sizeof(typename ContainerT::value_type);
+    if ((!targetResource && resource)
+        || (resource && targetResource && resource->is_equal(*targetResource))) {
+        auto message = resource->getMessage(static_cast<void *>(
+            const_cast<typename std::remove_const<typename ContainerT::value_type>::type *>(
+                container.data())));
+        if (message)
+            message->SetUsedSize(containerSizeBytes);
+        return std::move(message);
+    } else {
+        auto message = targetResource->getTransportFactory()->CreateMessage(containerSizeBytes);
+        std::memcpy(static_cast<fair::mq::byte *>(message->GetData()),
+                    container.data(),
+                    containerSizeBytes);
+        return std::move(message);
+    }
+};
+
+//_________________________________________________________________________________________________
+/// Return a vector of const ElemT, no wonership transfer.
+/// Resource must be kept alive throughout the lifetime of the
+/// container and associated message.
+template<typename ElemT>
+std::vector<const ElemT, boost::container::pmr::polymorphic_allocator<const ElemT>> getVector(
+    size_t nelem,
+    SpectatorMessageResource *resource)
+{
+    return std::vector<const ElemT, SpectatorAllocator<const ElemT>>(
```

Comment:
Why is the conversion from `std::vector<const ElemT, SpectatorAllocator<const ElemT>>` to `std::vector<const ElemT, boost::container::pmr::polymorphic_allocator<const ElemT>>` possible? At least https://en.cppreference.com/w/cpp/container/vector/vector doesn't list any ctor that would allow it.

Reply:
that is because SpectatorAllocator derives from polymorphic_allocator i guess...

Reply:
That doesn't help. `std::vector` still doesn't have a ctor that takes a `std::vector` with different allocator type: https://godbolt.org/z/1bGCWi

Reply:
hmm, somehow here it works... actually, in the original implementation in O2 the return type is auto, I had to invent something here for c++11. It is removed from this PR anyway, it lives on in #110 - but it's days are probably numbered as it is a bit hard to use - the owning version is better (or less bad ;) )

Reply:
ah i think I know why it worked here: it never did - the template is never instantiated as the function is never used :)

Reply:
:-) I suspected as much. Normally, going from one allocator type to the next implies a reallocating and moving the data. That's why I believe this conversion ctor is not part of the `std::vector` interface. ...and why `polymorphic_allocator` was needed so badly.